### PR TITLE
[inductor] Reduce excessive view materialization

### LINKED ISSUE #595: Inductor excessively materializes views in backward
Repro in https://gist.github.com/ngimel/887241bb5d7bb34adae4f9a329dcfea6 (includes python source, AOT graphs and compiled program), originally encountered in hf_Bert model. Repro doesn't generate exact same code as full model generates, but is enough to reproduce a problem. 
Repro is nn.Linear layer followed by an activation (GeLU or ReLU, ReLU generates simpler code but still has this problem). This is an excessively common pattern. In eager with ReLU activation, non-gemm kernels take approx 200 us on A100, inductor generated non-gemm kernels are 400 us, in large part due to excessive view materialization. 
In backward, gradInput for the activation is used to compute gradients of bias, weight and input. AOT graph shows that for weight and input gradient computation, a view or permute op is performed that doesn't involve data movement. Inductor generates code that copies activation gradInput (twice) instead of taking a view of it. 
I don't quite know how to fix it, so advice welcome.

### PR DESCRIPTION
**Pull request by jansel**
Fixes #595

For this example from @ngimel:
https://gist.github.com/ngimel/1a7156a98a5bebab31e15f5b5f6b222d

```diff
--- before	2022-08-03 14:26:38.528766870 -0700
+++ after	2022-08-03 14:48:11.040013800 -0700
@@ -46,12 +46,11 @@
 call_function  permute_default_6   aten.permute.default     (permute_default, [1, 0])                                                                                {}
 output         output              output                   ([view_default_3, view_default, view_default_1, view_default_2, permute_default_2, permute_default_6],)  {}
 INFO torchinductor.scheduler: RUN EXTERN buf0
-INFO torchinductor.scheduler: blocked names: {MemoryDep(name='buf0', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf1'))], MemoryDep(name='buf1', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf2')), SchedulerNodeBox(value=SchedulerNode(name='buf5'))], StarDep(name='buf2'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf3'))], MemoryDep(name='buf3', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf4'))]} deps: {'buf0': [SchedulerNodeBox(value=SchedulerNode(name='buf1'))], 'buf1': [SchedulerNodeBox(value=SchedulerNode(name='buf2')), SchedulerNodeBox(value=SchedulerNode(name='buf5'))], 'buf2': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf3'))], 'buf3': [SchedulerNodeBox(value=SchedulerNode(name='buf4'))]}
+INFO torchinductor.scheduler: blocked names: {MemoryDep(name='buf0', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf1'))], MemoryDep(name='buf1', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf2'))], StarDep(name='buf2'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf3'))], MemoryDep(name='buf3', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf4'))]} deps: {'buf0': [SchedulerNodeBox(value=SchedulerNode(name='buf1'))], 'buf1': [SchedulerNodeBox(value=SchedulerNode(name='buf2'))], 'buf2': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf3'))], 'buf3': [SchedulerNodeBox(value=SchedulerNode(name='buf4'))]}
 INFO torchinductor.scheduler: new fusable_deps: set()
 INFO torchinductor.scheduler: RUN buf1
 INFO torchinductor.scheduler: RUN buf2
-INFO torchinductor.scheduler: RUN buf5
-INFO torchinductor.codegen.triton: codegen numel=6291456 reduction_numel=1 nodes=3
+INFO torchinductor.codegen.triton: codegen numel=6291456 reduction_numel=1 nodes=2
 INFO torchinductor.scheduler: NEW KERNEL
 INFO torchinductor.scheduler: blocked names: {StarDep(name='buf2'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf3'))], MemoryDep(name='buf3', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf4'))]} deps: {'buf2': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf3'))], 'buf3': [SchedulerNodeBox(value=SchedulerNode(name='buf4'))]}
 INFO torchinductor.scheduler: new fusable_deps: set()
@@ -79,14 +78,13 @@
 
 @pointwise_heuristics(size_hints=[8388608])
 @triton.jit
-def kernel0(in_ptr0, in_ptr1, out_ptr0, out_ptr1, out_ptr2, xnumel, XBLOCK : tl.constexpr):
+def kernel0(in_ptr0, in_ptr1, out_ptr0, out_ptr1, xnumel, XBLOCK : tl.constexpr):
     xoffset = tl.program_id(0) * XBLOCK
     xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK])
     xmask = xindex < xnumel
-    x0 = xindex % 768
-    x1 = (xindex // 768)
     x2 = xindex
-    tmp0 = tl.load(in_ptr0 + x0 + (768*x1), xmask).to(tl.float32)
+    x0 = xindex % 768
+    tmp0 = tl.load(in_ptr0 + x2, xmask).to(tl.float32)
     tmp1 = tl.load(in_ptr1 + x0, xmask).to(tl.float32)
     tmp2 = tmp0 + tmp1
     tmp3 = 0.5
@@ -128,9 +126,8 @@
     tmp39 = tmp12 * tmp38
     tmp40 = tmp39 + tmp13
     tmp41 = tmp4 * tmp40
-    tl.store(out_ptr0 + x0 + (768*x1), tmp2, xmask)
+    tl.store(out_ptr0 + x2, tmp2, xmask)
     tl.store(out_ptr1 + x2, tmp41, xmask)
-    tl.store(out_ptr2 + x2, tmp41, xmask)
 
 @pointwise_heuristics(size_hints=[8388608])
 @triton.jit
@@ -155,14 +152,13 @@
     buf0 = empty_strided((8192, 768), (768, 1), device='cuda', dtype=torch.float16)
     aten.mm.out(as_strided(primals_5, (8192, 768), (768, 1)), as_strided(primals_2, (768, 768), (1, 768)), out=buf0)
     buf1 = empty_strided((8192, 768), (768, 1), device='cuda', dtype=torch.float16)
-    buf2 = empty_strided((8192, 768), (768, 1), device='cuda', dtype=torch.float16)
-    buf5 = empty_strided((8192, 768), (768, 1), device='cuda', dtype=torch.float16)
-    kernel0[grid(6291456)](buf0, primals_1, buf1, buf2, buf5, 6291456)
+    buf2 = empty_strided((16, 512, 768), (393216, 768, 1), device='cuda', dtype=torch.float16)
+    kernel0[grid(6291456)](buf0, primals_1, buf1, buf2, 6291456)
     buf3 = buf0; del buf0  # reuse
-    aten.mm.out(buf2, as_strided(primals_4, (768, 768), (1, 768)), out=buf3)
-    buf4 = as_strided(buf2, (16, 512, 768), (393216, 768, 1)); del buf2  # reuse
+    aten.mm.out(as_strided(buf2, (8192, 768), (768, 1)), as_strided(primals_4, (768, 768), (1, 768)), out=buf3)
+    buf4 = empty_strided((8192, 768), (768, 1), device='cuda', dtype=torch.float16)
     kernel1[grid(6291456)](buf3, primals_3, buf4, 6291456)
-    return (buf4, as_strided(primals_5, (8192, 768), (768, 1)), as_strided(buf1, (16, 512, 768), (393216, 768, 1)), buf5, as_strided(primals_4, (768, 768), (768, 1)), as_strided(primals_2, (768, 768), (768, 1)), )
+    return (as_strided(buf4, (16, 512, 768), (393216, 768, 1)), as_strided(primals_5, (8192, 768), (768, 1)), as_strided(buf1, (16, 512, 768), (393216, 768, 1)), as_strided(buf2, (8192, 768), (768, 1)), as_strided(primals_4, (768, 768), (768, 1)), as_strided(primals_2, (768, 768), (768, 1)), )
 
 
 if __name__ == "__main__":
@@ -175,7 +171,7 @@
     primals_5 = rand_strided((16, 512, 768), (393216, 768, 1), device='cuda', dtype=torch.float16)
     print_performance(lambda: call(primals_1, primals_2, primals_3, primals_4, primals_5))
 
-PyCodeCache /tmp/torchinductor_jansel/sy/csy7cydkea2lnbgx3nj2rxkfoddwhutlejfv4ovmf6itpmujw6d6.py
+PyCodeCache /tmp/torchinductor_jansel/si/csipw6zzgadfyw6agrvuwirfmv5rvljure7ir3tdspqinbv3alus.py
 BACKWARD GRAPH:
 opcode         name                  target                   args                                                                                       kwargs
 -------------  --------------------  -----------------------  -----------------------------------------------------------------------------------------  --------
@@ -239,23 +235,20 @@
 call_function  view_default_9        aten.view.default        (mm_default_4, [16, 512, 768])                                                             {}
 call_function  permute_default_9     aten.permute.default     (permute_default_8, [1, 0])                                                                {}
 output         output                output                   ([view_default_8, permute_default_9, view_default_5, permute_default_5, view_default_9],)  {}
-DEBUG torchinductor.ir: convert_to_reinterpret_view failed: stride=[768, 1] offset=0 index=d1 + 768*ModularIndexing(d0, 1, 512) + 393216*ModularIndexing(d0, 512, 16)
 INFO torchinductor.scheduler: RUN EXTERN buf0
-INFO torchinductor.scheduler: blocked names: {MemoryDep(name='buf0', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf3'))], MemoryDep(name='buf3', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf4')), SchedulerNodeBox(value=SchedulerNode(name='buf6')), SchedulerNodeBox(value=SchedulerNode(name='buf8'))], StarDep(name='buf4'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5'))], StarDep(name='buf6'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf7'))]} deps: {'buf0': [SchedulerNodeBox(value=SchedulerNode(name='buf3'))], 'buf3': [SchedulerNodeBox(value=SchedulerNode(name='buf4')), SchedulerNodeBox(value=SchedulerNode(name='buf6')), SchedulerNodeBox(value=SchedulerNode(name='buf8'))], 'buf4': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5'))], 'buf6': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf7'))]}
+INFO torchinductor.scheduler: blocked names: {MemoryDep(name='buf0', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf3'))], StarDep(name='buf3'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf4')), SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5'))], MemoryDep(name='buf3', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf6'))]} deps: {'buf0': [SchedulerNodeBox(value=SchedulerNode(name='buf3'))], 'buf3': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf4')), SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5')), SchedulerNodeBox(value=SchedulerNode(name='buf6'))]}
 INFO torchinductor.scheduler: new fusable_deps: set()
 INFO torchinductor.scheduler: RUN EXTERN buf1
-INFO torchinductor.scheduler: blocked names: {MemoryDep(name='buf3', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf4')), SchedulerNodeBox(value=SchedulerNode(name='buf6')), SchedulerNodeBox(value=SchedulerNode(name='buf8'))], StarDep(name='buf4'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5'))], StarDep(name='buf6'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf7'))]} deps: {'buf3': [SchedulerNodeBox(value=SchedulerNode(name='buf4')), SchedulerNodeBox(value=SchedulerNode(name='buf6')), SchedulerNodeBox(value=SchedulerNode(name='buf8'))], 'buf4': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5'))], 'buf6': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf7'))]}
+INFO torchinductor.scheduler: blocked names: {StarDep(name='buf3'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf4')), SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5'))], MemoryDep(name='buf3', index=c0, size=(6291456,)): [SchedulerNodeBox(value=SchedulerNode(name='buf6'))]} deps: {'buf3': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf4')), SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5')), SchedulerNodeBox(value=SchedulerNode(name='buf6'))]}
 INFO torchinductor.scheduler: new fusable_deps: set()
 INFO torchinductor.scheduler: RUN buf3
-INFO torchinductor.scheduler: RUN buf4
-INFO torchinductor.scheduler: RUN buf6
 INFO torchinductor.scheduler: RUN buf2
-INFO torchinductor.scheduler: RUN buf8
-INFO torchinductor.codegen.triton: codegen numel=768 reduction_numel=8192 nodes=5
+INFO torchinductor.scheduler: RUN buf6
+INFO torchinductor.codegen.triton: codegen numel=768 reduction_numel=8192 nodes=3
 INFO torchinductor.scheduler: NEW KERNEL
-INFO torchinductor.scheduler: blocked names: {StarDep(name='buf4'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5'))], StarDep(name='buf6'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf7'))]} deps: {'buf4': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5'))], 'buf6': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf7'))]}
+INFO torchinductor.scheduler: blocked names: {StarDep(name='buf3'): [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf4')), SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5'))]} deps: {'buf3': [SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf4')), SchedulerNodeBox(value=ExternKernelSchedulerNode(name='buf5')), SchedulerNodeBox(value=None)]}
 INFO torchinductor.scheduler: new fusable_deps: set()
-INFO torchinductor.scheduler: RUN EXTERN buf7
+INFO torchinductor.scheduler: RUN EXTERN buf4
 INFO torchinductor.scheduler: RUN EXTERN buf5
 
 from ctypes import c_void_p, c_long
@@ -275,7 +268,7 @@
 
 @reduction_heuristics(size_hints=[1024, 8192])
 @triton.jit
-def kernel0(in_ptr0, in_ptr1, in_ptr2, out_ptr1, out_ptr2, out_ptr3, out_ptr4, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
+def kernel0(in_ptr0, in_ptr1, in_ptr2, out_ptr0, out_ptr1, out_ptr2, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
     xoffset = tl.program_id(0) * XBLOCK
     xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK, 1])
     xmask = xindex < xnumel
@@ -342,12 +335,11 @@
         tmp51 = tmp50.to(tl.float32)
         _tmp53 = tl.where(xmask & rmask, _tmp53 + tmp52, _tmp53)
         _tmp54 = tl.where(xmask & rmask, _tmp54 + tmp51, _tmp54)
-        tl.store(out_ptr1 + r1 + (8192*x0), tmp51, xmask & rmask)
-        tl.store(out_ptr2 + r1 + (8192*x0), tmp51, xmask & rmask)
+        tl.store(out_ptr0 + r1 + (8192*x0), tmp51, xmask & rmask)
     tmp53 = tl.reshape(tl.sum(_tmp53, 1), [XBLOCK, 1])
-    tl.store(out_ptr3 + x0, tmp53, xmask)
+    tl.store(out_ptr1 + x0, tmp53, xmask)
     tmp54 = tl.reshape(tl.sum(_tmp54, 1), [XBLOCK, 1])
-    tl.store(out_ptr4 + x0, tmp54, xmask)
+    tl.store(out_ptr2 + x0, tmp54, xmask)
 
 
 def call(view_default, view_default_1, view_default_2, permute_default_2, permute_default_6, tangents_1):
@@ -361,17 +353,15 @@
     aten.mm.out(as_strided(tangents_1, (8192, 768), (768, 1)), permute_default_2, out=buf0)
     buf1 = empty_strided((768, 768), (768, 1), device='cuda', dtype=torch.float16)
     aten.mm.out(as_strided(tangents_1, (768, 8192), (1, 768)), view_default_2, out=buf1)
-    buf4 = empty_strided((8192, 768), (768, 1), device='cuda', dtype=torch.float16)
-    buf6 = empty_strided((768, 8192), (1, 768), device='cuda', dtype=torch.float16)
+    buf3 = empty_strided((16, 512, 768), (393216, 768, 1), device='cuda', dtype=torch.float16)
     buf2 = empty_strided((1, 768), (768, 1), device='cuda', dtype=torch.float16)
-    buf8 = empty_strided((1, 768), (768, 1), device='cuda', dtype=torch.float16)
-    kernel0[grid(768)](buf0, view_default_1, tangents_1, buf4, buf6, buf2, buf8, 768, 8192)
-    del buf0
-    buf7 = empty_strided((768, 768), (768, 1), device='cuda', dtype=torch.float16)
-    aten.mm.out(buf6, view_default, out=buf7)
-    buf5 = as_strided(buf6, (8192, 768), (768, 1)); del buf6  # reuse
-    aten.mm.out(buf4, permute_default_6, out=buf5)
-    return (as_strided(buf8, (768, ), (1, )), as_strided(buf7, (768, 768), (768, 1)), as_strided(buf2, (768, ), (1, )), as_strided(buf1, (768, 768), (768, 1)), as_strided(buf5, (16, 512, 768), (393216, 768, 1)), )
+    buf6 = empty_strided((1, 768), (768, 1), device='cuda', dtype=torch.float16)
+    kernel0[grid(768)](buf0, view_default_1, tangents_1, buf3, buf2, buf6, 768, 8192)
+    buf4 = buf0; del buf0  # reuse
+    aten.mm.out(as_strided(buf3, (8192, 768), (768, 1)), permute_default_6, out=buf4)
+    buf5 = empty_strided((768, 768), (768, 1), device='cuda', dtype=torch.float16)
+    aten.mm.out(as_strided(buf3, (768, 8192), (1, 768)), view_default, out=buf5)
+    return (as_strided(buf6, (768, ), (1, )), as_strided(buf5, (768, 768), (768, 1)), as_strided(buf2, (768, ), (1, )), as_strided(buf1, (768, 768), (768, 1)), as_strided(buf4, (16, 512, 768), (393216, 768, 1)), )
 
 
 if __name__ == "__main__":
@@ -385,4 +375,4 @@
     tangents_1 = rand_strided((16, 512, 768), (393216, 768, 1), device='cuda', dtype=torch.float16)
     print_performance(lambda: call(view_default, view_default_1, view_default_2, permute_default_2, permute_default_6, tangents_1))
 
-PyCodeCache /tmp/torchinductor_jansel/n5/cn5zq5tvftqf4njo4c46zt7br4cs557ftsph7n27arp7qhhjgdd6.py
+PyCodeCache /tmp/torchinductor_jansel/nu/cnu5mp3zgkpszf7bc3xkusewrabsrimb6unei5gvmukkmivpssmx.py

```

### CODE REVIEW

File: torchinductor/ir.py [hunk 1]
```diff
@@ -1876,15 +1884,18 @@ def convert_to_reinterpret_view(cls, x):
         if isinstance(x, ReinterpretView):
             return x
 
-        x.data.freeze_layout()
+        x.unwrap().freeze_layout()
         rw = extract_read_writes(x.make_loader(), x.get_size(), normalize=False)
         assert len(rw.reads) == 1
 
-        index = V.graph.sizevars.simplify(list(rw.reads)[0].index)
+        index = V.graph.sizevars.simplify_with_ranges(
```
Comment (ngimel) [inline torchinductor/ir.py:1891]: Yep, simplify_with_ranges has a better chance of producing a simpler expr

**Review by ngimel — APPROVED** [review]

File: torchinductor/ir.py [hunk 2]
```diff
@@ -758,8 +758,11 @@ def create(cls, x, *, dim=None):
             )
             return ReinterpretView(storage, new_layout)
 
-        # redirect to a generic view
-        return View.create(x, [s for s in x.get_size() if s != 1])
+        if dim is None:
+            # redirect to a generic view
+            return View.create(x, [s for s in x.get_size() if s != 1])
+        else:
+            return View.create(x, [s for i, s in enumerate(x.get_size()) if i != dim])
```
Comment (ngimel) [inline torchinductor/ir.py:765]: need also to assert that that size(dim) is 1?

Comment (jansel) [reply torchinductor/ir.py:765]: Thanks I'll add that.

**Comment by jansel** [comment]
This PR is causing a `RuntimeError: CUDA error: an illegal memory access was encountered` in hf_GPT2 fp16 training, but otherwise is working.  I don't think it is related, but this could be uncovering another bug.

**Comment by jansel** [comment]
@ngimel the stack trace I am getting is:
```
$ CUDA_LAUNCH_BLOCKING=1 ./benchmarks/torchbench.py -dcuda --inductor --training --float16 -k hf_GPT2
...
Traceback (most recent call last):
  File "/home/jansel/torchdynamo/benchmarks/common.py", line 879, in run_one_model
    new_result = model_iter_fn(model, example_inputs)
  File "./benchmarks/torchbench.py", line 362, in forward_and_backward_pass
    def forward_and_backward_pass(self, mod, inputs, collect_outputs=True):
  File "./benchmarks/torchbench.py", line 362, in forward_and_backward_pass
    def forward_and_backward_pass(self, mod, inputs, collect_outputs=True):
  File "./benchmarks/torchbench.py", line 368, in forward_and_backward_pass
    self.grad_scaler.scale(loss).backward()
  File "/home/jansel/functorch/functorch/_src/monkey_patching.py", line 77, in _backward
    return _old_backward(*args, **kwargs)
  File "/home/jansel/pytorch/torch/_tensor.py", line 371, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/jansel/pytorch/torch/autograd/__init__.py", line 191, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: CUDA error: an illegal memory access was encountered
```

It doesn't even look like the generated Triton code is in the call stack.  Is there a good way to figure out what kernel is causing the issue?

**Comment by ngimel** [comment]
Try `TORCH_SHOW_CPP_STACKTRACES=1 CUDA_LAUNCH_BLOCKING=1`, python stack traces are not really informative

**Comment by jansel** [comment]
```
$ TORCH_SHOW_CPP_STACKTRACES=1 CUDA_LAUNCH_BLOCKING=1 ./benchmarks/torchbench.py -dcuda --inductor --training -k hf_GPT2
cuda train hf_GPT2                            WARNING:root:using triton random, expect difference from eager
ERROR:root:unhandled error
Traceback (most recent call last):
  File "/home/jansel/torchdynamo/benchmarks/common.py", line 879, in run_one_model
    new_result = model_iter_fn(model, example_inputs)
  File "./benchmarks/torchbench.py", line 362, in forward_and_backward_pass
    def forward_and_backward_pass(self, mod, inputs, collect_outputs=True):
  File "./benchmarks/torchbench.py", line 362, in forward_and_backward_pass
    def forward_and_backward_pass(self, mod, inputs, collect_outputs=True):
  File "./benchmarks/torchbench.py", line 368, in forward_and_backward_pass
    self.grad_scaler.scale(loss).backward()
  File "/home/jansel/functorch/functorch/_src/monkey_patching.py", line 77, in _backward
    return _old_backward(*args, **kwargs)
  File "/home/jansel/pytorch/torch/_tensor.py", line 371, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/jansel/pytorch/torch/autograd/__init__.py", line 191, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: CUDA error: an illegal memory access was encountered
Exception raised from launch_vectorized_kernel at /home/jansel/pytorch/aten/src/ATen/native/cuda/CUDALoops.cuh:99 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6c (0x7f8f64316d4c in /home/jansel/pytorch/torch/lib/libc10.so)
frame #1: void at::native::gpu_kernel_impl<at::native::CUDAFunctor_add<float> >(at::TensorIteratorBase&, at::native::CUDAFunctor_add<float> const&) + 0xfa4 (0x7f8f674d66a4 in /home/jansel/pytorch/torch/lib/libtorch_cuda.so)
frame #2: void at::native::gpu_kernel<at::native::CUDAFunctor_add<float> >(at::TensorIteratorBase&, at::native::CUDAFunctor_add<float> const&) + 0x36b (0x7f8f674d71fb in /home/jansel/pytorch/torch/lib/libtorch_cuda.so)
frame #3: <unknown function> + 0x30d5ab8 (0x7f8f67480ab8 in /home/jansel/pytorch/torch/lib/libtorch_cuda.so)
frame #4: at::native::add_kernel(at::TensorIteratorBase&, c10::Scalar const&) + 0x34 (0x7f8f674823a4 in /home/jansel/pytorch/torch/lib/libtorch_cuda.so)
frame #5: <unknown function> + 0x2ed5fe7 (0x7f8f67280fe7 in /home/jansel/pytorch/torch/lib/libtorch_cuda.so)
frame #6: <unknown function> + 0x2ed60cb (0x7f8f672810cb in /home/jansel/pytorch/torch/lib/libtorch_cuda.so)
frame #7: at::_ops::add_Tensor::redispatch(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, c10::Scalar const&) + 0x8f (0x7f8f6fe55d7f in /home/jansel/pytorch/torch/lib/libtorch_cpu.so)
frame #8: <unknown function> + 0x398170f (0x7f8f7162870f in /home/jansel/pytorch/torch/lib/libtorch_cpu.so)
frame #9: <unknown function> + 0x3981d6e (0x7f8f71628d6e in /home/jansel/pytorch/torch/lib/libtorch_cpu.so)
frame #10: at::_ops::add_Tensor::call(at::Tensor const&, at::Tensor const&, c10::Scalar const&) + 0x17d (0x7f8f6fe8206d in /home/jansel/pytorch/torch/lib/libtorch_cpu.so)
frame #11: <unknown function> + 0x41c3494 (0x7f8f71e6a494 in /home/jansel/pytorch/torch/lib/libtorch_cpu.so)
frame #12: <unknown function> + 0x41c443e (0x7f8f71e6b43e in /home/jansel/pytorch/torch/lib/libtorch_cpu.so)
frame #13: torch::autograd::Engine::evaluate_function(std::shared_ptr<torch::autograd::GraphTask>&, torch::autograd::Node*, torch::autograd::InputBuffer&, std::shared_ptr<torch::autograd::ReadyQueue> const&) + 0x1004 (0x7f8f71e55f54 in /home/jansel/pytorch/torch/lib/libtorch_cpu.so)
frame #14: torch::autograd::Engine::thread_main(std::shared_ptr<torch::autograd::GraphTask> const&) + 0x651 (0x7f8f71e56fd1 in /home/jansel/pytorch/torch/lib/libtorch_cpu.so)
frame #15: torch::autograd::Engine::thread_init(int, std::shared_ptr<torch::autograd::ReadyQueue> const&, bool) + 0x9f (0x7f8f71e4e06f in /home/jansel/pytorch/torch/lib/libtorch_cpu.so)
frame #16: torch::autograd::python::PythonEngine::thread_init(int, std::shared_ptr<torch::autograd::ReadyQueue> const&, bool) + 0x67 (0x7f8f7c3769f7 in /home/jansel/pytorch/torch/lib/libtorch_python.so)
frame #17: <unknown function> + 0xd6de4 (0x7f8f7e4cdde4 in /lib/x86_64-linux-gnu/libstdc++.so.6)
frame #18: <unknown function> + 0x8609 (0x7f8fb3420609 in /lib/x86_64-linux-gnu/libpthread.so.0)
frame #19: clone + 0x43 (0x7f8fb3345133 in /lib/x86_64-linux-gnu/libc.so.6)
```

Hrm, looks like `at::native::add_kernel`.

Are we somehow creating an invalid tensor?

**Comment by ngimel** [comment]
hrm, that's really weird. Can you try disabling cuda graphs? We've had a few cases where cudagraphs are inexplicably causing errors #601

**Comment by jansel** [comment]
> hrm, that's really weird. Can you try disabling cuda graphs? We've had a few cases where cudagraphs are inexplicably causing errors #601

Same thing without cudagraphs...

**Comment by ngimel** [comment]
It executes all the forward and backward triton kernels, and likely the error happens when the grads are accumulated by pytorch itself (hence the add operation) (ugh, we really should be tracing and doing it ourselves within AOT, but that's for next time). I'll try to narrow down what's going wrong, maybe some tensor memory gets deallocated before it is sent to the accumulation op.

**Comment by Chillee** [comment]
@ngimel I'm not sure we want to trace down the accumulations in AOTAutograd 🤔

**Comment by jansel** [comment]
I just ran the rest of huggingface benchmarks (that fit in 24GB ram), and none of them see this same issue other than hf_GPT2.  (Lots of other HF fixes in https://github.com/pytorch/torchdynamo/pull/728 from that effort.)

Given this seems to only effect hf_GPT2, I'd propose temporarily skipping that one and landing this.

**Comment by ngimel** [comment]
[Here](https://gist.github.com/ngimel/f4850eaa4a5403e6094a85dd38ab6a48) is generated backward for hf_gpt2. If I run with cuda graphs, the error is with gradient accumulation corresponding to output #172, which, in turn, is `constant1` in the generated backward kernel. It's weird that `constant` is returned in gradient? cc @chillee
I wanted to print `constant1` within the generated compiled program,  but I need to disable cudagraphs for this, and if I disable cudagraphs, I'm getting the error (also illegal memory access) in the 7th kernel in backward.

**Comment by Chillee** [comment]
@ngimel I mean... it's not impossible? Could you print out the AOTAutograd FX graph instead? (or try running it with aotautograd-nop backend).

**Comment by ngimel** [comment]
Gist can't handle the graphs, so here's the paste [P520959570](https://www.internalfb.com/intern/paste/P520959570/)

**Comment by Chillee** [comment]
@ngimel the AOTAutograd graph also returns a constant for that value, so seems right to me.

**Comment by ngimel** [comment]
It's very weird, by looking at the graph it seems like corresponding input is embedding weight, so its gradient shouldn't be constant?

**Comment by Chillee** [comment]
@ngimel It's ... possible it could be some weird thing with in-place modification of a tensor constant, caused by recursive decompositions without re-entrant autograd?

**Pull request merged**